operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1673 @@
+	}
+}
+
+func (ps *PoolStore) Upsert(pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) {
+	if pool == nil {
+		return
+	}
+	ps.pools[pool.GetUID()] = pool
+}
+
+func (ps *PoolStore) Delete(pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) {
+	delete(ps.pools, pool.GetUID())
+}
+
+func (ps *PoolStore) GetByUID(uid types.UID) (*cilium_api_v2alpha1.CiliumLoadBalancerIPPool, bool) {
+	pool, found := ps.pools[uid]
+	return pool, found
+}
+
+type RangesStore struct {
+	ranges       []*LBRange
+	poolToRanges map[types.UID][]*LBRange
+}
+
+func NewRangesStore() RangesStore {
+	return RangesStore{
+		poolToRanges: make(map[types.UID][]*LBRange),
+	}
+}
+
+func (rs *RangesStore) Delete(lbRange *LBRange) {
+	idx := slices.Index(rs.ranges, lbRange)
+	if idx != -1 {
+		rs.ranges = slices.Delete(rs.ranges, idx, idx+1)
+	}
+
+	poolRanges := rs.poolToRanges[lbRange.originPool]
+
+	idx = slices.Index(poolRanges, lbRange)
+	if idx != -1 {
+		poolRanges = slices.Delete(poolRanges, idx, idx+1)
+	}
+
+	if len(poolRanges) > 0 {
+		rs.poolToRanges[lbRange.originPool] = poolRanges
+	} else {
+		delete(rs.poolToRanges, lbRange.originPool)
+	}
+}
+
+func (rs *RangesStore) Add(lbRange *LBRange) {
+	rs.ranges = append(rs.ranges, lbRange)
+	poolRanges := rs.poolToRanges[lbRange.originPool]
+	poolRanges = append(poolRanges, lbRange)
+	rs.poolToRanges[lbRange.originPool] = poolRanges
+}
+
+func (rs *RangesStore) GetRangesForPool(uid types.UID) ([]*LBRange, bool) {
+	ranges, found := rs.poolToRanges[uid]
+	return ranges, found
+}
+
+type LBRange struct {
+	// the actual data of which ips have been allocated or not
+	allocRange *ipallocator.Range
+	// If true, the LB range has been disabled via the CRD and thus no IPs should be allocated from this range
+	externallyDisabled bool
+	// If true, the LB range has been disabled by us, because it conflicts with other ranges for example.
+	// This range should not be used for allocation.
+	internallyDisabled bool
+	// The UID of the pool that originated this LB range
+	originPool types.UID
+}
+
+func NewLBRange(cidr *net.IPNet, pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) (*LBRange, error) {
+	allocRange, err := ipallocator.NewCIDRRange(cidr)
+	if err != nil {
+		return nil, fmt.Errorf("new cidr range: %w", err)
+	}
+
+	return &LBRange{
+		allocRange:         allocRange,
+		internallyDisabled: false,
+		externallyDisabled: pool.Spec.Disabled,
+		originPool:         pool.GetUID(),
+	}, nil
+}
+
+func (lr *LBRange) Disabled() bool {
+	return lr.internallyDisabled || lr.externallyDisabled
+}
+
+func (lr *LBRange) String() string {
+	cidr := lr.allocRange.CIDR()
+	return fmt.Sprintf(
+		"%s (free: %d, used: %d, intDis: %v, extDis: %v) - origin %s",
+		cidr.String(),
+		lr.allocRange.Free(),
+		lr.allocRange.Used(),
+		lr.internallyDisabled,
+		lr.externallyDisabled,
+		lr.originPool,
+	)
+}
+
+func ipNetStr(net net.IPNet) string {
+	ptr := &net
+	return ptr.String()
+}
+
+type ServiceStore struct {
```

Comment:
This should be in its own file and should be its own cell with its own interface type and set of tests.